### PR TITLE
Bug/41

### LINKED ISSUE #41: In Windows, Using the installer with the Minecraft Launcher open can lead to silent install failure
If a default Client install is done while the Minecraft launcher is open, it could lead to the profile not being created.

If after performing a default client install with the Minecraft Launcher open, the Minecraft Launcher is restarted, the profile created by the Fabric installer will show up in the list of Installations.

Any changes to the configuration of any of the "Installations" done in the Minecraft Launcher after performing a default client install, or creating a new "Installation" manually on the Minecraft Launcher will overwrite the changes to launcher\_profiles.json done by the Fabric Installer.

It seems the Minecraft launcher writes the changes on exit (for simple changes) and immediately when creating a new profile, in both cases overwriting the changes made by the Fabric Installer.

Another problem is that a user may expect that the profile created by the Fabric installer will show up the the Installations list automatically without the need to restart the Minecraft Launcher.

solution 1: The Fabric installer should detect that the Minecraft Launcher is open and warn the user to close it before proceeding. 

solution 2: A simple message box warning the user to close the Minecraft Launcher, without performing any detection.

### PR DESCRIPTION
**Pull request by DeathsGun**
This closes #41 by executing a command to list all running processes:
Under windows it will use `tasklist.exe` and under linux and mac os it will use `ps`.
It will open a dialog to warn the user about the opened launcher.

### CODE REVIEW

**Comment by liach** [comment]
Don't think this is a good solution for 3rd-party launchers like MultiMC.

**Comment by asiekierka** [comment]
Isn't fabric-installer meant to be only used with the vanilla launcher, though?

File: src/main/java/net/fabricmc/installer/util/Utils.java [hunk 1]
```diff
@@ -31,109 +33,131 @@
...
+
+    public static void launcherCheck(boolean showGui) throws IOException {
+        List<String> processes = getRunningProcesses();
+        if (processes.stream().anyMatch(process -> process.toLowerCase().contains("minecraft"))) {
+            System.out.println("The Minecraft launcher is open! Please close it before installing fabric!");
+            if (showGui) {
+                JOptionPane.showMessageDialog(null,
+                        "The Minecraft launcher is open! Please close it before installing fabric!");
+            }
+            System.exit(0);
```
Comment (liach) [inline src/main/java/net/fabricmc/installer/util/Utils.java:149]: Imo this should fail than exit with 0

File: src/main/java/net/fabricmc/installer/Main.java [hunk 2]
```diff
@@ -43,18 +40,23 @@
...
 					JOptionPane.showMessageDialog(null, "You are using an outdated version of Java, Fabric will not work! Please update to Java 8 or newer to use Fabric.", "Java Version Warning", JOptionPane.ERROR_MESSAGE);
 				}
 				return;
 			}
 		}
-
+		try {
+			Utils.launcherCheck(showGui);
+		} catch (IOException e) {
+			e.printStackTrace();
```
Comment (liach) [inline src/main/java/net/fabricmc/installer/Main.java:58]: Maybe show a screen asking users to make sure minecraft launcher is closed if our fallback commands failed. (of course check showGui first)

Comment (DeathsGun) [reply src/main/java/net/fabricmc/installer/Main.java:58]: Is a message box ok? Or should it be a confirmation dialog?

Comment (liach) [reply src/main/java/net/fabricmc/installer/Main.java:58]: Hmm, do you think we should ask user to check about launcher status if we cannot obtain the process information? It's up to you how you handle this case. But just know that check may fail due to process info being inaccessible.

**Review by liach — APPROVED** [review]

**Comment by modmuss50** [comment]
I think this could be handled better in the native bootstrap.

**Comment by sfPlayer1** [comment]
Bootstrap should do as little as possible, this PR however is too generic, the problem shouldn't affect a version only install.

**Comment by modmuss50** [comment]
I do want to look to see if we can use the windows [CreateMutex API](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createmutexa) (or similar) to detect if the mojang launcher is open, it should be the most reliable method to do this.

With the mojang launcher open I can see the following mutant:

**Comment by DeathsGun** [comment]
But then it will only work on Windows

**Comment by modmuss50** [comment]
Yes, would need to explore other options for the other systems. But using this is better than searching the process list IMO.

It will need a bit of work to get this into java however, not too sure on the best approach to that.

**Comment by modmuss50** [comment]
I have found a solution for MacOS as well, we can use https://developer.apple.com/documentation/appkit/nsworkspace/1534059-runningapplications to list all the running applications and find if the mc launcher's bundle id is there.

Proof of concept. 

Screenshot 2021-05-08 at 11 36 04

**Comment by DeathsGun** [comment]
I close this as the merge request in the native part was merged and this isn't needed anymore

**Pull request closed without merging**